open-craft/opencraft
# Load Balancer, Instant Switchover & SSL Support

This implements support for load-balancing servers.

For background on the configuration of the load-balancing server, please see https://github.com/open-craft/ansible-load-balancer/blob/master/README.md.

If you prefer, you can review the first commit separately.  It is a self-contained refactoring that doesn't change any functionality.

The selection algorithm to select a load-balancing server is very simple at the moment; it chooses one at random.  For now this should be enough, since we only have a single server.
### Testing instructions
1. Add the following settings to the `.env` file on your vagrant instance:
   
   ```
   LOAD_BALANCER_FRAGMENT_NAME_PREFIX='$USERNAME-'
   PRELIMINARY_PAGE_SERVER_IP='149.202.186.237'
   ```
   
   `$USERNAME` is some name identifying you, and is used to recognise your configuration file fragments on the dev load-balancing server, which is shared between all of us.
2. Run `make migrate`.
3. Add the public SSH key of your vagrant instance (probably `/home/vagrant/.ssh/id_rsa.pub`) to the authrorised keys for the `ubuntu` user (`/home/ubuntu/.ssh/authorized_keys`) on the server `haproxy-dev.opencraft.hosting`.  You can use the "ssh key for databases instances" from the secrets repository to log into that server.
4. Test that you can log into `haproxy-dev.opencraft.hosting` from your vagrant instance using `ssh ubuntu@haproxy-dev.opencraft.hosting`.
5. In the Django shell, run
   
   ```
   LoadBalancingServer.objects.create(
       domain="haproxy-dev.opencraft.hosting",
       ssh_username="ubuntu")
   ```
6. Create a new instance and spawn an appserver, e.g.
   
   ```
   instance = OpenEdXInstance.objects.create(
       sub_domain="$USERNAME-haproxy-test",
       name="haproxy test",
       configuration_source_repo_url="https://github.com/open-craft/configuration.git",
       configuration_version="opencraft-release/eucalyptus.2",
       edx_platform_repository_url="https://github.com/open-craft/edx-platform.git",
       edx_platform_commit="opencraft-release/eucalyptus.2",
       openedx_release="open-release/eucalyptus.2",
   )
   from instance.tasks import spawn_appserver
   spawn_appserver(instance.ref.pk)
   ```
   
   The DNS record will be set up immediately, but will of course take some time to propagate.  Once the load-balancing server can verify that the DNS record points to it, it will request an SSL certificate for the new domain names.  Once this has happened, you should be able to access the link to the LMS in the instance manager, and it should show the page informing you that the instance is being provisioned _via HTTPS_.  (In the real production setup, we have a wildcard DNS entry for *.opencraft.hosting, so users will see the splash screen even before DNS propagation, via plain HTTP.)
7. Activate the appserver once it's fully deployed.  Since no DNS propagation is necessary to activate the appserver, it should become available under the domain name in less than one minute.
